### DISCUSSION TITLE
RP2040 Host Multiple bulk transfers failing

### DISCUSSION THREAD

**Discussion by ReFil** [hathach/tinyusb#2004] [Q&A]
I'm trying to use the RP2040 in host mode with the arduino tinyusb library to write several data packets in quick succession to a single endpoint using tuh_edpt_xfer(); The first transfer returns 1 but all subsequent transfers return 0. Reads from a different endpoint are succeeding so the communication is definitely there, the first transfer is 4 bytes and the subsequent transfers all fail. Do i need to reopen the endpoint after each transfer? I also think there's a blocking issue or race condition as i was having problems reading data until i added a small delay, but not sure where it could be happening. Any help would be much appreciated

my code looks like this 
` 
endpoint1_out_desc = {
    .bLength = 7,
    .bDescriptorType = TUSB_DESC_ENDPOINT,

    .bEndpointAddress = 0x01, // Endpoint address with OUT direction
    .bmAttributes = {.xfer = TUSB_XFER_BULK},
    .wMaxPacketSize = 64,
    .bInterval = 0
    };

uint8_t zeros1[800] = {0x00};

  Serial.println("Opening data out endpoint");
  tuh_edpt_open(DeviceAddress, &endpoint1_out_desc);  
  Serial.println("Starting Stage1 transfer");
  
  uint32_t stage1 = 0x000FFFFF;
  uint8_t *stage1ptr = (uint8_t *)&stage1;
  tuh_xfer_t transfer1 = {
      .daddr = DeviceAddress,
      .ep_addr = endpoint1_out_desc.bEndpointAddress, 
      .actual_len = 4,
      .buflen = 4,
      .buffer = stage1ptr,
      .complete_cb = NULL,
      .user_data = 0,
    };  
  Serial.println("Performing Stage1 transfer");
  bool success = tuh_edpt_xfer(&transfer1);
  delay(10);
  Serial.print("Transfer report: ");
  Serial.println(success);
  Serial.println("Starting Stage2 transfer");
  tuh_xfer_t transfer2 = {
      .daddr = tegraDeviceAddress,
      .ep_addr = endpoint1_out_desc.bEndpointAddress, 
      .actual_len = 800,
      .buflen = 800,
      .buffer = zeros1,
      .complete_cb = NULL,
    };
  
  Serial.println("Performing Stage2 transfer");
  success = tuh_edpt_xfer(&transfer2);
  delay(1000);
  Serial.print("Transfer report: ");
  Serial.println(success);`

**Comment by hathach** [comment]
have you tried to troubleshoot to see why it return false, enable LOG, add print() to the tuh_xfer()

**Reply by ReFil** [reply]
I originally tried attaching a debugger but my development environment wasn't playing ball. if i enable #define CFG_TUSB_DEBUG        2 i can connect a uart adapter to Serial1 (gpio 0 and gpio 1) right? Sorry for the basic question, i'm not 100% clear on how to use the built in logging functionality

**Reply by hathach** [reply]
bump up your tinyusb lib to latest, change the CFG_TUSB_DEBUG to 2 in tusb_config_rp2040.h. It should have debug message output to Serial1 on default tx/rx pin (check your variant.h)

**Reply by ReFil** [reply]
Sorry for the delay, i didnt have much time to test this, here's some relevant excerps from the logs
EP 81 with 16 bytes
tuh_task_ext 490: ASSERT FAILED
on EP 01 with 4 bytes
tuh_task_ext 490: ASSERT FAILED
on EP 00 with 8 bytes
120
on EP 00 with 0 bytes
[1:1] Control FAILED, xferred_bytes = 0

It seems like there's some conflict between USBD logging and USBH logging too, is there some way to only log USBH?

**Reply by ReFil** [reply]
I've put logging in tuh_edpt_xfer and discovered that the second transfer fails at the usbh_edpt_claim step and just loops there

  Open EP 01 with Size = 64
Verify endpointClaim edptEdpt good, init transfer  Queue EP 01 with 4 bytes ... OK
Verify endpointClaim edptVerify endpointClaim edptVerify endpointClaim edptVerify endpointClaim edptVerify endpoint

**Reply by ReFil** [reply]
I've tried a few things but haven't been able to make it work, it still cannot claim the endpoint from the previous transfer it seems
